File: container-di/src/test/scala/com/yahoo/container/di/ContainerTest.scala
// Copyright 2017 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.di

import com.yahoo.component.AbstractComponent
import com.yahoo.config.di.IntConfig
import com.yahoo.config.test.TestConfig
import com.yahoo.container.bundle.MockBundle
import com.yahoo.container.di.ContainerTest._
import com.yahoo.container.di.componentgraph.Provider
import com.yahoo.container.di.componentgraph.core.ComponentGraphTest.{SimpleComponent, SimpleComponent2}
import com.yahoo.container.di.componentgraph.core.ComponentNode.ComponentConstructorException
import com.yahoo.container.di.componentgraph.core.{ComponentGraph, Node}
import com.yahoo.container.di.config.RestApiContext
import org.hamcrest.CoreMatchers._
import org.junit.Assert._
import org.junit.{After, Before, Ignore, Test}

import scala.collection.JavaConverters._
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.duration._
import scala.concurrent.{Await, Future}
import scala.language.{existentials, postfixOps}
import scala.util.Try

/**
 * @author tonytv
 * @author gjoranv
 */
class ContainerTest {
  var dirConfigSource: DirConfigSource = _

  @Before def setup()  {
    dirConfigSource = new DirConfigSource("ContainerTest-")
  }

  @After def cleanup() {
    dirConfigSource.cleanup()
  }

  @Test
  def components_can_be_created_from_config() {
    writeBootstrapConfigs()
    dirConfigSource.writeConfig("test", """stringVal "myString" """)

    val container = newContainer(dirConfigSource)

    val component = createComponentTakingConfig(container.getNewConfigGraph())
    assertThat(component.config.stringVal(), is("myString"))

    container.shutdownConfigurer()
  }

  @Test
  def components_are_reconfigured_after_config_update_without_bootstrap_configs() {
    writeBootstrapConfigs()
    dirConfigSource.writeConfig("test", """stringVal "original" """)

    val container = newContainer(dirConfigSource)

    val componentGraph = container.getNewConfigGraph()
    val component = createComponentTakingConfig(componentGraph)

    assertThat(component.config.stringVal(), is("original"))

    // Reconfigure
    dirConfigSource.writeConfig("test", """stringVal "reconfigured" """)
    container.reloadConfig(2)

    val newComponentGraph = container.getNewConfigGraph(componentGraph)
    val component2 = createComponentTakingConfig(newComponentGraph)
    assertThat(component2.config.stringVal(), is("reconfigured"))

    container.shutdownConfigurer()
  }

  @Test
  def graph_is_updated_after_bootstrap_update() {
    dirConfigSource.writeConfig("test", """stringVal "original" """)
    writeBootstrapConfigs("id1")

    val container = newContainer(dirConfigSource)

    val graph = container.getNewConfigGraph()
    val component = createComponentTakingConfig(graph)
    assertThat(component.getId.toString, is("id1"))

    writeBootstrapConfigsWithMultipleComponents(Array(
      ("id1", classOf[ComponentTakingConfig]),
      ("id2", classOf[ComponentTakingConfig])))

    container.reloadConfig(2)
    val newGraph = container.getNewConfigGraph(graph)

    assertThat(ComponentGraph.getNode(newGraph, "id1"), notNullValue(classOf[Node]))
    assertThat(ComponentGraph.getNode(newGraph, "id2"), notNullValue(classOf[Node]))

    container.shutdownConfigurer()
  }

  //@Test TODO
  def deconstructor_is_given_guice_components() {
  }

  @Test
  def osgi_component_is_deconstructed_when_not_reused() {
    writeBootstrapConfigs("id1", classOf[DestructableComponent])

    val container = newContainer(dirConfigSource)

    val oldGraph = container.getNewConfigGraph()
    val componentToDestruct = oldGraph.getInstance(classOf[DestructableComponent])

    writeBootstrapConfigs("id2", classOf[DestructableComponent])
    container.reloadConfig(2)
    container.getNewConfigGraph(oldGraph)
    assertTrue(componentToDestruct.deconstructed)
  }

  @Ignore  // because logAndDie is impossible(?) to verify programmatically
  @Test
  def manually_verify_what_happens_when_first_graph_contains_component_that_throws_exception_in_ctor() {
    writeBootstrapConfigs("thrower", classOf[ComponentThrowingExceptionInConstructor])
    val container = newContainer(dirConfigSource)
    var currentGraph: ComponentGraph = null
    try {
      currentGraph = container.getNewConfigGraph()
      fail("Expected to log and die.")
    } catch {
      case _: Throwable => fail("Expected to log and die")
    }
  }

  @Test
  def previous_graph_is_retained_when_new_graph_contains_component_that_throws_exception_in_ctor() {
    val simpleComponentEntry = ComponentEntry("simpleComponent", classOf[SimpleComponent])

    writeBootstrapConfigs(Array(simpleComponentEntry))
    val container = newContainer(dirConfigSource)
    var currentGraph = container.getNewConfigGraph()

    val simpleComponent = currentGraph.getInstance(classOf[SimpleComponent])

    writeBootstrapConfigs("thrower", classOf[ComponentThrowingExceptionInConstructor])
    container.reloadConfig(2)
    try {
      currentGraph = container.getNewConfigGraph(currentGraph)
      fail("Expected exception")
    } catch {
      case _: ComponentConstructorException => // Expected, do nothing
      case _: Throwable => fail("Expected ComponentConstructorException")
    }
    assertEquals(1, currentGraph.generation)

    // Also verify that next reconfig is successful
    val componentTakingConfigEntry = ComponentEntry("componentTakingConfig", classOf[ComponentTakingConfig])
    dirConfigSource.writeConfig("test", """stringVal "myString" """)
    writeBootstrapConfigs(Array(simpleComponentEntry, componentTakingConfigEntry))
    container.reloadConfig(3)
    currentGraph = container.getNewConfigGraph(currentGraph)

    assertEquals(3, currentGraph.generation)
    assertSame(simpleComponent, currentGraph.getInstance(classOf[SimpleComponent]))
    assertNotNull(currentGraph.getInstance(classOf[ComponentTakingConfig]))
  }

  @Test
  def previous_graph_is_retained_when_new_graph_throws_exception_for_missing_config() {
    val simpleComponentEntry = ComponentEntry("simpleComponent", classOf[SimpleComponent])

    writeBootstrapConfigs(Array(simpleComponentEntry))
    val container = newContainer(dirConfigSource)
    var currentGraph = container.getNewConfigGraph()

    val simpleComponent = currentGraph.getInstance(classOf[SimpleComponent])

    writeBootstrapConfigs("thrower", classOf[ComponentThrowingExceptionForMissingConfig])
    dirConfigSource.writeConfig("test", """stringVal "myString" """)
    container.reloadConfig(2)
    try {
      currentGraph = container.getNewConfigGraph(currentGraph)
      fail("Expected exception")
    } catch {
      case _: IllegalArgumentException => // Expected, do nothing
      case _: Throwable => fail("Expected IllegalArgumentException")
    }
    assertEquals(1, currentGraph.generation)
  }

  @Test
  def runOnce_hangs_waiting_for_valid_config_after_invalid_config() {
    dirConfigSource.writeConfig("test", """stringVal "original" """)
    writeBootstrapConfigs("myId", classOf[ComponentTakingConfig])

    val container = newContainer(dirConfigSource)
    var currentGraph = container.getNewConfigGraph()

    writeBootstrapConfigs("thrower", classOf[ComponentThrowingExceptionForMissingConfig])
    container.reloadConfig(2)

    try {
      currentGraph = container.getNewConfigGraph(currentGraph)
      fail("expected exception")
    } catch {
      case e: Exception =>
    }

    val newGraph = Future {
      currentGraph = container.getNewConfigGraph(currentGraph)
      currentGraph
    }

    Try {
      Await.ready(newGraph, 1 second)
    } foreach { x => fail("Expected waiting for new config.") }


    writeBootstrapConfigs("myId2", classOf[ComponentTakingConfig])
    container.reloadConfig(3)

    assertNotNull(Await.result(newGraph, 5 minutes))
  }


  @Test
  def bundle_info_is_set_on_rest_api_context() {
    val clazz = classOf[RestApiContext]

    writeBootstrapConfigs("restApiContext", clazz)
    dirConfigSource.writeConfig("jersey-bundles", """bundles[0].spec "mock-entry-to-enforce-a-MockBundle" """)
    dirConfigSource.writeConfig("jersey-injection", """inject[0]" """)

    val container = newContainer(dirConfigSource)
    val componentGraph = container.getNewConfigGraph()

    val restApiContext = componentGraph.getInstance(clazz)
    assertNotNull(restApiContext)

    assertThat(restApiContext.getBundles.size, is(1))
    assertThat(restApiContext.getBundles.get(0).symbolicName, is(MockBundle.SymbolicName))
    assertThat(restApiContext.getBundles.get(0).version, is(MockBundle.BundleVersion))

    container.shutdownConfigurer()
   }

  @Test
  def restApiContext_has_all_components_injected() {
    new JerseyInjectionTest {
      assertFalse(restApiContext.getInjectableComponents.isEmpty)
      assertThat(restApiContext.getInjectableComponents.size(), is(2))

      container.shutdownConfigurer()
    }
  }

  // TODO: reuse injectedComponent as a named component when we support that
  trait JerseyInjectionTest {
    val restApiClass = classOf[RestApiContext]
    val injectedClass = classOf[SimpleComponent]
    val injectedComponentId = "injectedComponent"
    val anotherComponentClass = classOf[SimpleComponent2]
    val anotherComponentId = "anotherComponent"

    val componentsConfig: String =
      ComponentEntry(injectedComponentId, injectedClass).asConfig(0) + "\n" +
        ComponentEntry(anotherComponentId, anotherComponentClass).asConfig(1) + "\n" +
        ComponentEntry("restApiContext", restApiClass).asConfig(2) + "\n" +
        s"components[2].inject[0].id $injectedComponentId\n" +
        s"components[2].inject[1].id $anotherComponentId\n"

    val injectionConfig = s"""inject[1]
                             |inject[0].instance $injectedComponentId
                             |inject[0].forClass "${injectedClass.getName}"
                            """.stripMargin

    dirConfigSource.writeConfig("components", componentsConfig)
    dirConfigSource.writeConfig("bundles", "")
    dirConfigSource.writeConfig("jersey-bundles", """bundles[0].spec "mock-entry-to-enforce-a-MockBundle" """)
    dirConfigSource.writeConfig("jersey-injection", injectionConfig)

    val container = newContainer(dirConfigSource)
    val componentGraph = container.getNewConfigGraph()

    val restApiContext = componentGraph.getInstance(restApiClass)
  }

  case class ComponentEntry(componentId: String,  classId: Class[_]) {
    def asConfig(position: Int): String = {
      <config>
      |components[{position}].id "{componentId}"
      |components[{position}].classId "{classId.getName}"
      |components[{position}].configId "{dirConfigSource.configId}"
      </config>.text.stripMargin.trim
    }
  }

  def writeBootstrapConfigs(componentEntries: Array[ComponentEntry]) {
    dirConfigSource.writeConfig("bundles", "")
    dirConfigSource.writeConfig("components", """
      components[%s]
      %s
    """.format(componentEntries.length,
      componentEntries.zipWithIndex.map{ case (entry, index) => entry.asConfig(index) }.mkString("\n")))
  }

  def writeBootstrapConfigs(componentId: String = classOf[ComponentTakingConfig].getName,
                            classId: Class[_] = classOf[ComponentTakingConfig]) {

    writeBootstrapConfigs(Array(ComponentEntry(componentId, classId)))
  }

  def writeBootstrapConfigsWithMultipleComponents(idAndClass: Array[(String, Class[_])]) {
    writeBootstrapConfigs(idAndClass.map{case(id, classId) => ComponentEntry(id, classId)})
  }


  @Test
  def providers_are_destructed() {
    writeBootstrapConfigs("id1", classOf[DestructableProvider])

    val deconstructor = new ComponentDeconstructor {
      def deconstruct(component: AnyRef) {
        component match  {
          case c : AbstractComponent => c.deconstruct()
          case p : Provider[_] => p.deconstruct()
        }
      }
    }

    val container = newContainer(dirConfigSource, deconstructor)

    val oldGraph = container.getNewConfigGraph()
    val destructableEntity = oldGraph.getInstance(classOf[DestructableEntity])

    writeBootstrapConfigs("id2", classOf[DestructableProvider])
    container.reloadConfig(2)
    container.getNewConfigGraph(oldGraph)

    assertTrue(destructableEntity.deconstructed)
  }
}


object ContainerTest {
  class DestructableEntity {
    var deconstructed = false
  }

  class DestructableProvider extends Provider[DestructableEntity] {
    val instance = new DestructableEntity

    def get() = instance

    def deconstruct() {
      require(! instance.deconstructed)
      instance.deconstructed = true
    }
  }

  class ComponentTakingConfig(val config: TestConfig) extends AbstractComponent {
    require(config != null)
  }

  class ComponentThrowingExceptionInConstructor() {
    throw new RuntimeException("This component fails upon construction.")
  }

  class ComponentThrowingExceptionForMissingConfig(intConfig: IntConfig) extends AbstractComponent {
    fail("This component should never be created. Only used for tests where 'int' config is missing.")
  }

  class DestructableComponent extends AbstractComponent {
    var deconstructed = false
    override def deconstruct() {
      deconstructed = true
    }
  }

  class TestDeconstructor extends ComponentDeconstructor {
    def deconstruct(component: AnyRef) {
      component match {
        case vespaComponent: DestructableComponent => vespaComponent.deconstruct()
        case _ =>
      }
    }
  }

  private def newContainer(dirConfigSource: DirConfigSource,
                           deconstructor: ComponentDeconstructor = new TestDeconstructor()):
  Container = {
    new Container(new CloudSubscriberFactory(dirConfigSource.configSource), dirConfigSource.configId, deconstructor)
  }

  def createComponentTakingConfig(componentGraph: ComponentGraph): ComponentTakingConfig = {
    componentGraph.getInstance(classOf[ComponentTakingConfig])
  }

  def convertMap[K, V](map: java.util.Map[K, V]): Map[K, V] = map.asScala.toMap
}

File: container-di/src/test/java/com/yahoo/container/di/ContainerTest.java
// Copyright 2018 Yahoo Holdings. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
package com.yahoo.container.di;

import com.yahoo.component.AbstractComponent;
import com.yahoo.config.di.IntConfig;
import com.yahoo.config.test.TestConfig;
import com.yahoo.container.bundle.MockBundle;
import com.yahoo.container.di.componentgraph.Provider;
import com.yahoo.container.di.componentgraph.core.ComponentGraph;
import com.yahoo.container.di.componentgraph.core.ComponentGraphTest.SimpleComponent;
import com.yahoo.container.di.componentgraph.core.ComponentGraphTest.SimpleComponent2;
import com.yahoo.container.di.componentgraph.core.ComponentNode.ComponentConstructorException;
import com.yahoo.container.di.componentgraph.core.Node;
import com.yahoo.container.di.config.RestApiContext;
import org.junit.Ignore;
import org.junit.Test;

import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.TimeoutException;

import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.CoreMatchers.notNullValue;
import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertSame;
import static org.junit.Assert.assertThat;
import static org.junit.Assert.assertTrue;
import static org.junit.Assert.fail;

/**
 * @author Tony Vaagenes
 * @author gjoranv
 * @author ollivir
 */
public class ContainerTest extends ContainerTestBase {

    @Test
    public void components_can_be_created_from_config() {
        writeBootstrapConfigs();
        dirConfigSource.writeConfig("test", "stringVal \"myString\"");

        Container container = newContainer(dirConfigSource);

        ComponentTakingConfig component = createComponentTakingConfig(container.getNewComponentGraph());
        assertThat(component.config.stringVal(), is("myString"));

        container.shutdownConfigurer();
    }

    @Test
    public void components_are_reconfigured_after_config_update_without_bootstrap_configs() {
        writeBootstrapConfigs();
        dirConfigSource.writeConfig("test", "stringVal \"original\"");

        Container container = newContainer(dirConfigSource);

        ComponentGraph componentGraph = container.getNewComponentGraph();
        ComponentTakingConfig component = createComponentTakingConfig(componentGraph);

        assertThat(component.config.stringVal(), is("original"));

        // Reconfigure
        dirConfigSource.writeConfig("test", "stringVal \"reconfigured\"");
        container.reloadConfig(2);

        ComponentGraph newComponentGraph = container.getNewComponentGraph(componentGraph);
        ComponentTakingConfig component2 = createComponentTakingConfig(newComponentGraph);
        assertThat(component2.config.stringVal(), is("reconfigured"));

        container.shutdownConfigurer();
    }

    @Test
    public void graph_is_updated_after_bootstrap_update() {
        dirConfigSource.writeConfig("test", "stringVal \"original\"");
        writeBootstrapConfigs("id1");

        Container container = newContainer(dirConfigSource);

        ComponentGraph graph = container.getNewComponentGraph();
        ComponentTakingConfig component = createComponentTakingConfig(graph);
        assertThat(component.getId().toString(), is("id1"));

        writeBootstrapConfigs(
                new ComponentEntry("id1", ComponentTakingConfig.class),
                new ComponentEntry("id2", ComponentTakingConfig.class));

        container.reloadConfig(2);
        ComponentGraph newGraph = container.getNewComponentGraph(graph);

        assertThat(ComponentGraph.getNode(newGraph, "id1"), notNullValue(Node.class));
        assertThat(ComponentGraph.getNode(newGraph, "id2"), notNullValue(Node.class));

        container.shutdownConfigurer();
    }

    //@Test TODO
    public void deconstructor_is_given_guice_components() {
    }

    @Test
    public void osgi_component_is_deconstructed_when_not_reused() {
        writeBootstrapConfigs("id1", DestructableComponent.class);

        Container container = newContainer(dirConfigSource);

        ComponentGraph oldGraph = container.getNewComponentGraph();
        DestructableComponent componentToDestruct = oldGraph.getInstance(DestructableComponent.class);

        writeBootstrapConfigs("id2", DestructableComponent.class);
        container.reloadConfig(2);
        container.getNewComponentGraph(oldGraph);
        assertTrue(componentToDestruct.deconstructed);
    }

    @Ignore  // because logAndDie is impossible(?) to verify programmatically
    @Test
    public void manually_verify_what_happens_when_first_graph_contains_component_that_throws_exception_in_ctor() {
        writeBootstrapConfigs("thrower", ComponentThrowingExceptionInConstructor.class);
        Container container = newContainer(dirConfigSource);
        try {
            container.getNewComponentGraph();
            fail("Expected to log and die.");
        } catch (Throwable t) {
            fail("Expected to log and die");
        }
    }

    @Test
    public void previous_graph_is_retained_when_new_graph_contains_component_that_throws_exception_in_ctor() {
        ComponentEntry simpleComponentEntry = new ComponentEntry("simpleComponent", SimpleComponent.class);

        writeBootstrapConfigs(simpleComponentEntry);
        Container container = newContainer(dirConfigSource);
        ComponentGraph currentGraph = container.getNewComponentGraph();

        SimpleComponent simpleComponent = currentGraph.getInstance(SimpleComponent.class);

        writeBootstrapConfigs("thrower", ComponentThrowingExceptionInConstructor.class);
        container.reloadConfig(2);
        try {
            currentGraph = container.getNewComponentGraph(currentGraph);
            fail("Expected exception");
        } catch (ComponentConstructorException ignored) {
            // Expected, do nothing
        } catch (Throwable t) {
            fail("Expected ComponentConstructorException");
        }
        assertEquals(1, currentGraph.generation());

        // Also verify that next reconfig is successful
        ComponentEntry componentTakingConfigEntry = new ComponentEntry("componentTakingConfig", ComponentTakingConfig.class);
        dirConfigSource.writeConfig("test", "stringVal \"myString\"");
        writeBootstrapConfigs(simpleComponentEntry, componentTakingConfigEntry);
        container.reloadConfig(3);
        currentGraph = container.getNewComponentGraph(currentGraph);

        assertEquals(3, currentGraph.generation());
        assertSame(simpleComponent, currentGraph.getInstance(SimpleComponent.class));
        assertNotNull(currentGraph.getInstance(ComponentTakingConfig.class));
    }

    @Test
    public void previous_graph_is_retained_when_new_graph_throws_exception_for_missing_config() {
        ComponentEntry simpleComponentEntry = new ComponentEntry("simpleComponent", SimpleComponent.class);

        writeBootstrapConfigs(simpleComponentEntry);
        Container container = newContainer(dirConfigSource);
        ComponentGraph currentGraph = container.getNewComponentGraph();

        currentGraph.getInstance(SimpleComponent.class);

        writeBootstrapConfigs("thrower", ComponentThrowingExceptionForMissingConfig.class);
        dirConfigSource.writeConfig("test", "stringVal \"myString\"");
        container.reloadConfig(2);
        try {
            currentGraph = container.getNewComponentGraph(currentGraph);
            fail("Expected exception");
        } catch (IllegalArgumentException ignored) {
            // Expected, do nothing
        } catch (Throwable t) {
            fail("Expected IllegalArgumentException");
        }
        assertEquals(1, currentGraph.generation());
    }

    @Test
    public void runOnce_hangs_waiting_for_valid_config_after_invalid_config() throws InterruptedException, ExecutionException, TimeoutException {
        dirConfigSource.writeConfig("test", "stringVal \"original\"");
        writeBootstrapConfigs("myId", ComponentTakingConfig.class);

        Container container = newContainer(dirConfigSource);
        final ComponentGraph currentGraph = container.getNewComponentGraph();

        writeBootstrapConfigs("thrower", ComponentThrowingExceptionForMissingConfig.class);
        container.reloadConfig(2);

        try {
            container.getNewComponentGraph(currentGraph);
            fail("expected exception");
        } catch (Exception ignored) {
        }
        ExecutorService exec = Executors.newFixedThreadPool(1);
        Future<ComponentGraph> newGraph = exec.submit(() -> container.getNewComponentGraph(currentGraph));

        try {
            newGraph.get(1, TimeUnit.SECONDS);
            fail("Expected waiting for new config.");
        } catch (Exception ignored) {
            // expect to time out
        }

        writeBootstrapConfigs("myId2", ComponentTakingConfig.class);
        container.reloadConfig(3);

        assertNotNull(newGraph.get(5, TimeUnit.MINUTES));
    }


    @Test
    public void bundle_info_is_set_on_rest_api_context() {
        Class<RestApiContext> clazz = RestApiContext.class;

        writeBootstrapConfigs("restApiContext", clazz);
        dirConfigSource.writeConfig("jersey-bundles", "bundles[0].spec \"mock-entry-to-enforce-a-MockBundle\"");
        dirConfigSource.writeConfig("jersey-injection", "inject[0]");

        Container container = newContainer(dirConfigSource);
        ComponentGraph componentGraph = container.getNewComponentGraph();

        RestApiContext restApiContext = componentGraph.getInstance(clazz);
        assertNotNull(restApiContext);

        assertThat(restApiContext.getBundles().size(), is(1));
        assertThat(restApiContext.getBundles().get(0).symbolicName, is(MockBundle.SymbolicName));
        assertThat(restApiContext.getBundles().get(0).version, is(MockBundle.BundleVersion));

        container.shutdownConfigurer();
    }

    @Test
    public void restApiContext_has_all_components_injected() {
        Class<RestApiContext> restApiClass = RestApiContext.class;
        Class<SimpleComponent> injectedClass = SimpleComponent.class;
        String injectedComponentId = "injectedComponent";
        Class<SimpleComponent2> anotherComponentClass = SimpleComponent2.class;
        String anotherComponentId = "anotherComponent";

        String componentsConfig =
                new ComponentEntry(injectedComponentId, injectedClass).asConfig(0) + "\n" +
                        new ComponentEntry(anotherComponentId, anotherComponentClass).asConfig(1) + "\n" +
                        new ComponentEntry("restApiContext", restApiClass).asConfig(2) + "\n" +
                        "components[2].inject[0].id " + injectedComponentId + "\n" +
                        "components[2].inject[1].id " + anotherComponentId + "\n";

        String injectionConfig = "inject[1]\n" +//
                "inject[0].instance " + injectedComponentId + "\n" +//
                "inject[0].forClass \"" + injectedClass.getName() + "\"\n";

        dirConfigSource.writeConfig("components", componentsConfig);
        dirConfigSource.writeConfig("bundles", "");
        dirConfigSource.writeConfig("jersey-bundles", "bundles[0].spec \"mock-entry-to-enforce-a-MockBundle\"");
        dirConfigSource.writeConfig("jersey-injection", injectionConfig);

        Container container = newContainer(dirConfigSource);
        ComponentGraph componentGraph = container.getNewComponentGraph();

        RestApiContext restApiContext = componentGraph.getInstance(restApiClass);

        assertFalse(restApiContext.getInjectableComponents().isEmpty());
        assertThat(restApiContext.getInjectableComponents().size(), is(2));

        container.shutdownConfigurer();
    }

    @Test
    public void providers_are_destructed() {
        writeBootstrapConfigs("id1", DestructableProvider.class);

        ComponentDeconstructor deconstructor = new ComponentDeconstructor() {
            @Override
            public void deconstruct(Object component) {
                if (component instanceof AbstractComponent) {
                    ((AbstractComponent) component).deconstruct();
                    ;
                } else if (component instanceof Provider) {
                    ((Provider<?>) component).deconstruct();
                }
            }
        };

        Container container = newContainer(dirConfigSource, deconstructor);

        ComponentGraph oldGraph = container.getNewComponentGraph();
        DestructableEntity destructableEntity = oldGraph.getInstance(DestructableEntity.class);

        writeBootstrapConfigs("id2", DestructableProvider.class);
        container.reloadConfig(2);
        container.getNewComponentGraph(oldGraph);

        assertTrue(destructableEntity.deconstructed);
    }

    static class DestructableEntity {
        private boolean deconstructed = false;
    }

    public static class DestructableProvider implements Provider<DestructableEntity> {
        DestructableEntity instance = new DestructableEntity();

        public DestructableEntity get() {
            return instance;
        }

        public void deconstruct() {
            assertFalse(instance.deconstructed);
            instance.deconstructed = true;
        }
    }

    public static class ComponentTakingConfig extends AbstractComponent {
        private final TestConfig config;

        public ComponentTakingConfig(TestConfig config) {
            assertNotNull(config);
            this.config = config;
        }
    }

    public static class ComponentThrowingExceptionInConstructor {
        public ComponentThrowingExceptionInConstructor() {
            throw new RuntimeException("This component fails upon construction.");
        }
    }

    public static class ComponentThrowingExceptionForMissingConfig extends AbstractComponent {
        public ComponentThrowingExceptionForMissingConfig(IntConfig intConfig) {
            fail("This component should never be created. Only used for tests where 'int' config is missing.");
        }
    }

    public static class DestructableComponent extends AbstractComponent {
        private boolean deconstructed = false;

        @Override
        public void deconstruct() {
            deconstructed = true;
        }
    }

    public static class TestDeconstructor implements ComponentDeconstructor {
        @Override
        public void deconstruct(Object component) {
            if (component instanceof DestructableComponent) {
                DestructableComponent vespaComponent = (DestructableComponent) component;
                vespaComponent.deconstruct();
            }
        }
    }

    private static Container newContainer(DirConfigSource dirConfigSource,
                                          ComponentDeconstructor deconstructor) {
        return new Container(new CloudSubscriberFactory(dirConfigSource.configSource), dirConfigSource.configId(), deconstructor);
    }

    private static Container newContainer(DirConfigSource dirConfigSource) {
        return newContainer(dirConfigSource, new TestDeconstructor());
    }

    private ComponentTakingConfig createComponentTakingConfig(ComponentGraph componentGraph) {
        return componentGraph.getInstance(ComponentTakingConfig.class);
    }
}
